Implement the raqote backend

Eijebong opened the pull request:
Rebased version of #23601 with some more stuff implemented. This passes some (not many) wpt tests.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/23936)
<!-- Reviewable:end -->

pylbrecht committed:
Implement DrawOptions.set_alpha()

pylbrecht committed:
Implement StrokeOptions

pylbrecht committed:
Make GenericPathBuilder take &mut self

pylbrecht committed:
Implement provided raqote::PathBuilder's methods for GenericPathBuilder

pylbrecht committed:
WIP: Make GenericPathBuilder's methods take &mut self

jdm committed:
Deal with fallout from mutable trait method change.

pylbrecht committed:
Implement fill() for raqote::DrawTarget

pylbrecht committed:
Implement some of GenericDrawTarget's methods for raqote

Eijebong committed:
Add a way to convert a FillOrStrokeStyle to a raqote Source

For now most of it unimplemented but it'll allow for some more testing

Eijebong committed:
Implement set_fill_style

Eijebong committed:
Implement Pattern::is_zero_size_gradient

There are probably other times where this would be true but I'll worry
about that later

Eijebong committed:
Add an implementation for fill_rect

Again, this is probably incomplete but it'll do for now

Eijebong committed:
Implement snapshot_data

Eijebong committed:
Implement need_to_draw_shadow

Eijebong committed:
Implement size_from_pattern

I tried to mimic the azure implementation but raqote doesn't have a
concept of horizontal or vertical repeat so this might be
wrong/incomplete but this is the last function needed to finally have
some tests pass

Eijebong committed:
Update raqote

Eijebong committed:
Make tidy happy

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

Eijebong commented on Cargo.toml:
```diff
@@ -28,3 +28,4 @@ opt-level = 3
 # Those are here to dedupe winapi since mio is still using winapi 0.2.
 mio = { git = "https://github.com/servo/mio.git", branch = "servo" }
 iovec = { git = "https://github.com/servo/iovec.git", branch = "servo" }
+
```
:scissors:

Eijebong commented on components/canvas/raqote_backend.rs:
```diff
@@ -72,47 +80,90 @@ impl Backend for RaqoteBackend {
     pub fn set_line_join(&mut self, _val: LineJoinStyle) {
-        unimplemented!()
+        match self {
+            StrokeOptions::Raqote(options, _) => options.join = _val.to_raqote_style(),
```
Rename to val

Eijebong commented on components/canvas/raqote_backend.rs:
```diff
@@ -72,47 +80,90 @@ impl Backend for RaqoteBackend {
     pub fn set_line_cap(&mut self, _val: LineCapStyle) {
-        unimplemented!()
+        match self {
+            StrokeOptions::Raqote(options, _) => options.cap = _val.to_raqote_style(),
```
Rename to val too

Eijebong commented on components/canvas/raqote_backend.rs:
```diff
@@ -72,47 +80,90 @@ impl Backend for RaqoteBackend {
     pub fn set_alpha(&mut self, _val: f32) {
         match self {
-            DrawOptions::Raqote(()) => unimplemented!(),
+            DrawOptions::Raqote(draw_options) => draw_options.alpha = _val,
```
val here too

Eijebong commented on components/canvas/raqote_backend.rs:
```diff
@@ -194,69 +247,242 @@ impl GenericDrawTarget for raqote::DrawTarget {
+            self,
+            &pb.finish(),
+            pattern.as_raqote(),
+            &raqote::DrawOptions::new(),
```
I think I could pass `draw_options.as_raqote()` here

Eijebong commented on components/canvas/raqote_backend.rs:
```diff
@@ -194,69 +247,242 @@ impl GenericDrawTarget for raqote::DrawTarget {
+}
+
+// TODO(pylbrecht)
+#[cfg(feature = "raqote_backend")]
```
:scissors: This is not needed anymore

Eijebong committed:
Implement clear_rect

jdm commented:
Some new compilation failures when building with the azure backend:
```
[0m[1m[38;5;9merror[E0726][0m[0m[1m: implicit elided lifetime not allowed here[0m
[0m   [0m[0m[1m[38;5;12m--> [0m[0mcomponents/canvas/azure_backend.rs:472:6[0m
[0m    [0m[0m[1m[38;5;12m|[0m
[0m[1m[38;5;12m472[0m[0m [0m[0m[1m[38;5;12m| [0m[0mimpl Pattern {[0m
[0m    [0m[0m[1m[38;5;12m| [0m[0m     [0m[0m[1m[38;5;9m^^^^^^^[0m[0m[1m[38;5;12m-[0m[0m [0m[0m[1m[38;5;12mhelp: indicate the anonymous lifetime: `<'_>`[0m

[0m[1m[38;5;9merror[E0726][0m[0m[1m: implicit elided lifetime not allowed here[0m
[0m   [0m[0m[1m[38;5;12m--> [0m[0mcomponents/canvas/azure_backend.rs:726:6[0m
[0m    [0m[0m[1m[38;5;12m|[0m
[0m[1m[38;5;12m726[0m[0m [0m[0m[1m[38;5;12m| [0m[0mimpl Pattern {[0m
[0m    [0m[0m[1m[38;5;12m| [0m[0m     [0m[0m[1m[38;5;9m^^^^^^^[0m[0m[1m[38;5;12m-[0m[0m [0m[0m[1m[38;5;12mhelp: indicate the anonymous lifetime: `<'_>`[0m

[0m[1m[38;5;9merror[E0392][0m[0m[1m: parameter `'a` is never used[0m
[0m   [0m[0m[1m[38;5;12m--> [0m[0mcomponents/canvas/canvas_data.rs:355:18[0m
[0m    [0m[0m[1m[38;5;12m|[0m
[0m[1m[38;5;12m355[0m[0m [0m[0m[1m[38;5;12m| [0m[0mpub enum Pattern<'a> {[0m
[0m    [0m[0m[1m[38;5;12m| [0m[0m                 [0m[0m[1m[38;5;9m^^[0m[0m [0m[0m[1m[38;5;9munused parameter[0m
[0m    [0m[0m[1m[38;5;12m|[0m
[0m    [0m[0m[1m[38;5;12m= [0m[0m[1mhelp[0m[0m: consider removing `'a` or using a marker such as `std::marker::PhantomData`[0m

[0m[1m[38;5;9merror[0m[0m[1m: aborting due to 3 previous errors[0m
```

Eijebong committed:
Implement stroke_line and set_stroke_style

Eijebong committed:
Properly pass draw_options in fill_rect

Eijebong committed:
Implement get_size

Eijebong committed:
Implement snapshot_data_owned

Eijebong committed:
Pass BlendMode::Clear to the fill method of DrawTarget in clear_rect

This makes some tests regarding clearing rects pass. Otherwise the rect
wouldn't get cleared properly and it'd keep its original color but get
an alpha of 0

Eijebong committed:
Set the line cap to butt when drawing a rect with a 0 width/height

Otherwise raqote will draw the cap even though it shouldn't because the
spec says so

Eijebong committed:
Match the azure backend regarding cap stuff

I think this is wrong because it won't respect cap style at all when
stroking lines but we'll cross that bridge when we get to it

Eijebong committed:
Implement set_shadow_color, set_global_composition and stroke_rect

Eijebong committed:
Imlplement get_composition_op

Eijebong committed:
Implement create_similar_draw_target

Eijebong committed:
Make tidy happy again

Eijebong committed:
WIP: Fix azure backend

Eijebong commented:
It's now sitting there:
```
Ran 830 tests finished in 421.0 seconds.
  • 516 ran as expected. 13 tests skipped.
  • 195 tests crashed unexpectedly
  • 1 tests failed unexpectedly
  • 57 tests timed out unexpectedly
  • 61 tests had unexpected subtest results
```

The azure backend fix needs https://github.com/servo/rust-azure/pull/302 because Path needs to be cloned but is otherwise complete (ran all tests with the patch and they passed)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #23950) made this pull request unmergeable. Please resolve the merge conflicts.

Eijebong commented:
Now at:
```
Ran 830 tests finished in 296.0 seconds.
  • 568 ran as expected. 13 tests skipped.
  • 172 tests crashed unexpectedly
  • 1 tests failed unexpectedly
  • 16 tests timed out unexpectedly
  • 73 tests had unexpected subtest results
```

Eijebong committed:
Implement a few functions to make tests in draw-image and draw-path pass

Eijebong committed:
Implement get_format

Since raqote doesn't have multiple formats for its surfaces, this is a
noop to avoid crashing with the unimplemented

jdm commented:
I'd love to merge these changes. Would it take much to address the various nits and rebase this PR?

Eijebong committed:
Implement Source::Surface for FillOrStrokeStyle with raqote

Eijebong committed:
Fix the build and tidy and address nits

Eijebong commented:
@jdm there

jdm commented:
@bors-servo r+
Thanks!

bors-servo commented:
:pushpin: Commit f829eae has been approved by `jdm`

<!-- @bors-servo r=jdm f829eaea03c3c84e773798cffc29186f7e4c007d -->

bors-servo commented:
:hourglass: Testing commit f829eaea03c3c84e773798cffc29186f7e4c007d with merge 0c7b71eb596d2c9340641ba2a3a6a758fcf71cc0...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/Oa18djNsT867vzrsprlgTQ)

Eijebong commented:
> [taskcluster:error] Task aborted - max run time exceeded

jdm commented:
@bors-servo retry
* killed the rogue test processes on mac5

bors-servo commented:
:hourglass: Testing commit f829eaea03c3c84e773798cffc29186f7e4c007d with merge 326336253be6cf58afba1ed788997f3084f69efa...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/13173)

Eijebong commented:
```
{
    "status": "CRASH", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": null, 
    "test": "/referrer-policy/generic/iframe-src-change.html", 
    "line": 192884, 
    "action": "test_result", 
    "expected": "OK"
}
```

jdm commented:
@bors-servo retry

bors-servo commented:
:hourglass: Testing commit f829eaea03c3c84e773798cffc29186f7e4c007d with merge d8a1c1e02ed97a1e5f13bb4c21e4f3b1bdbd4ad1...

bors-servo commented:
:sunny: Test successful - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/13176), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/13237), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/OwRmmwOVQIGVRwupZq3p3Q)
Approved by: jdm
Pushing d8a1c1e02ed97a1e5f13bb4c21e4f3b1bdbd4ad1 to master...
